src/google/cloud/ndb/model.py

```diff
@@ -523,8 +523,43 @@ def _entity_from_ds_entity(ds_entity, model_class=None):
     entity = model_class()
     if ds_entity.key:
         entity._key = key_module.Key._from_ds_key(ds_entity.key)
+
     for name, value in ds_entity.items():
         prop = getattr(model_class, name, None)
+
+        # Backwards compatibility shim. NDB previously stored structured
+        # properties as sets of dotted name proprties. Datastore now has native
+        # support for embedded entities and NDB now uses that, by default. This
+        # handles the case of reading structured properties from older
+        # NDB datastore instances.
+        if prop is None and "." in name:
+            supername, subname = name.split(".", 1)
+            structprop = getattr(model_class, supername, None)
+            if isinstance(structprop, StructuredProperty):
+                subvalue = value
+                value = structprop._get_base_value(entity)
+                if value in (None, []):  # empty list for repeated props
+                    kind = structprop._model_class._get_kind()
+                    key = key_module.Key(kind, None)
+                    if structprop._repeated:
+                        value = [
+                            _BaseValue(entity_module.Entity(key._key))
+                            for _ in subvalue
+                        ]
+                    else:
+                        value = entity_module.Entity(key._key)
+                        value = _BaseValue(value)
+
+                    structprop._store_value(entity, value)
+
+                if structprop._repeated:
+                    for subentity, subsubvalue in zip(value, subvalue):
+                        subentity.b_val.update({subname: subsubvalue})
+                else:
+                    value.b_val.update({subname: subvalue})
+
+                continue
```

Comment:
I've had trouble with coverage in continue statements before. Seems like the python optimizer just skips that and goes to the start of the loop directly, so the line is never reached. Solutions are: 1) Use "# pragma: no branch". 2) Turn off the optimizer. I have gone with 1) when I encounter this problem.